language/move-binary-format/src/compatibility.rs

```diff
@@ -128,14 +124,18 @@ impl Compatibility {
         // we can remove/change a friend function if the function is not used by any module in the
         // friend list. But for simplicity, we decided to go to the more restrictive form now and
         // we may revisit this in the future.
-        for (name, old_func) in &old_module.exposed_functions {
-            let new_func = match new_module.exposed_functions.get(name) {
+        for (name, old_func) in &old_module.functions {
+            let new_func = match new_module.functions.get(name) {
                 Some(new_func) => new_func,
                 None => {
                     if matches!(old_func.visibility, Visibility::Friend) {
                         friend_linking = false;
-                    } else {
-                        struct_and_pub_function_linking = false;
+                    } else if old_func.visibility != Visibility::Private {
+                        struct_and_function_linking = false;
+                    } else if old_func.is_entry && self.check_private_entry_linking {
+                        // This must be a private entry function. So set the link breakage if we're
+                        // checking for that.
+                        private_entry_linking = false;
                     }
                     continue;
                 }
```

Comment:
nit, `let ... else`?  Also, some consistency nits:

 - use `matches!` to check equivalence against `Visibility::Friend` but `==` on `Visibility::Private` -- might as well use equality on both.
 - checking for `self.check_private_entry_linking` -- it seems like the pattern for all the other checks is to do this once, at the end, so we might as well conform to that here?

```suggestion
            let Some(new_func) = new_module.functions.get(name) else {
               /* body of None match arm */
            }
```

Or even perhaps: 

```suggestion
            let Some(new_func) = new_module.functions.get(name) 
            else if old_func.visibility == Visibility::Friend {
                friend_linking = false;
                continue
            } else if old_func.visibility != Visibility::Private {
                struct_and_function_linking = false;
                continue
            } else if old_func.is_entry {
                private_entry_linking = false;
                continue
            }
```

Not sure if Rust allows that syntax...

Reply:
Sadly Rust syntax doesn't allow this :'( But moving the if-else into the `else { .. }` of the let-else looks OK (or, at least better than it does now).